File: backend/src/processors/analyzer.js
import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const RAW_DATA_PATH = path.resolve(__dirname, '../../data/raw');

/**
 * Analyze a dataset file from backend/data/raw/.
 * Adapt this function to the specific hackathon theme.
 */
export async function analyzeDataset(filename, analysisType = 'summary') {
  const filePath = path.join(RAW_DATA_PATH, filename);

  if (!fs.existsSync(filePath)) {
    throw new Error(`Dataset not found: ${filename}`);
  }

  const ext = path.extname(filename).toLowerCase();

  if (ext === '.json') {
    return analyzeJson(filePath, analysisType);
  }

  if (ext === '.csv') {
    return analyzeCsv(filePath, analysisType);
  }

  throw new Error(`Unsupported file type: ${ext}`);
}

async function analyzeJson(filePath, analysisType) {
  const raw = fs.readFileSync(filePath, 'utf8');
  const data = JSON.parse(raw);
  const records = Array.isArray(data) ? data : [data];

  return {
    format: 'json',
    rowCount: records.length,
    columns: records.length > 0 ? Object.keys(records[0]) : [],
    sample: analysisType === 'detailed' ? records.slice(0, 5) : undefined,
    insights: [],
  };
}

async function analyzeCsv(filePath, analysisType) {
  const { default: csvParser } = await import('csv-parser');
  const records = [];

  await new Promise((resolve, reject) => {
    fs.createReadStream(filePath)
      .pipe(csvParser())
      .on('data', (row) => records.push(row))
      .on('end', resolve)
      .on('error', reject);
  });

  return {
    format: 'csv',
    rowCount: records.length,
    columns: records.length > 0 ? Object.keys(records[0]) : [],
    sample: analysisType === 'detailed' ? records.slice(0, 5) : undefined,
    insights: [],
  };
}

File: backend/app/processors/analyzer.py
import pandas as pd
from pathlib import Path

# Resolves to data/raw/ at repo root regardless of where uvicorn is launched
RAW_DATA_PATH = Path(__file__).parents[4] / "data" / "raw"


def analyze_dataset(filename: str, analysis_type: str = "summary") -> dict:
    """
    Analyze a dataset file from data/raw/.
    Replace this function body with theme-specific logic on hackathon day.
    """
    file_path = RAW_DATA_PATH / filename

    if not file_path.exists():
        raise FileNotFoundError(f"Dataset not found in data/raw/: {filename}")

    ext = file_path.suffix.lower()

    if ext == ".csv":
        df = pd.read_csv(file_path)
    elif ext == ".json":
        df = pd.read_json(file_path)
    elif ext in (".xlsx", ".xls"):
        df = pd.read_excel(file_path)
    else:
        raise ValueError(f"Unsupported file type: {ext}")

    return _build_result(df, ext.lstrip("."), analysis_type)


def _build_result(df: pd.DataFrame, fmt: str, analysis_type: str) -> dict:
    result = {
        "format": fmt,
        "row_count": len(df),
        "columns": list(df.columns),
        "statistics": df.describe(include="all").fillna("").to_dict() if analysis_type == "detailed" else {},
        "sample": df.head(5).to_dict(orient="records") if analysis_type == "detailed" else [],
        "insights": [],
    }
    return result
